File: day-13/main.rb
# Advent of Code 2015 - Day 13

def read_input
  File.read(File.expand_path("../input.txt", __FILE__))
end

class AmityMap
  SEATING_REGEXP =
    /^(.+) would (gain|lose) (\d+) happiness units by sitting next to (.+)\.$/

  attr_reader :names
  attr_reader :amity

  def initialize
    @names = []
    @amity = []
  end

  def add_amity(line)
    foo, sign, happiness, bar = line.match(SEATING_REGEXP)[1..4]
    foo_index = name_index(foo)
    bar_index = name_index(bar)
    happiness = happiness.to_i * ((sign == "gain") ? +1 : -1)

    add_pair(foo_index, bar_index, happiness)
  end

  def add_self
    @normalized = false
    self_index = name_index("Self")
    (0...self_index).each do |other_index|
      add_pair(self_index, other_index, 0)
      add_pair(other_index, self_index, 0)
    end

    self
  end

  def max_happiness
    normalize_amity

    all_happiness = []
    each_perm_of_pairs do |pairs|
      all_happiness << pairs.map do |foo, bar|
        amity[foo][bar] + amity[bar][foo]
      end.reduce(0, :+)
    end
    all_happiness.max
  end

  private

  def add_pair(foo, bar, happiness)
    amity[foo] ||= []
    amity[foo][bar] = happiness
  end

  def each_perm_of_pairs
    upper_bound = names.length - 1
    (0..upper_bound).to_a.permutation do |perm|
      pairs = perm.each_cons(2).to_a + [[perm.last, perm.first]]
      yield pairs
    end
  end

  def name_index(name)
    names.index(name) || (names.push(name).length - 1)
  end

  def normalize_amity
    return unless @normalized
    @normalized = true

    upper_bound = names.length - 1
    (0..upper_bound).each do |index|
      amity[index] ||= []
      (0..upper_bound).each do |second_index|
        amity[index][second_index] ||= 0
      end
    end
  end
end

def main
  map = AmityMap.new

  lines = read_input.split("\n").reject(&:empty?)
  lines.each { |line| map.add_amity(line) }

  puts "Part One: #{map.max_happiness}"
  puts "Part Two: #{map.add_self.max_happiness}"
end

main if "main" == to_s

File: day-13/main.py
"""Advent of Code 2015 - Day 13."""

import itertools
import re
from pathlib import Path


def read_input():
    """Read input file and split into individual lines returned as a list."""
    return Path(__file__).with_name('input.txt').read_text().splitlines()


class AmityMap:
    """Map of people and their mutual amities (not symmetric)."""

    # Regular expression for parsing amity pairs from the input file.
    SEATING_REGEXP = re.compile(r'^(?P<foo>.+) would (?P<sign>gain|lose) '
                                r'(?P<happiness>\d+) happiness units by '
                                r'sitting next to (?P<bar>.+)\.$')

    @classmethod
    def from_lines(cls, lines):
        """Initialize a map from lines read from the input file."""
        amity_map = cls()
        for line in lines:
            match = cls.SEATING_REGEXP.search(line)
            happiness = int(match.group('happiness'))
            if match.group('sign') == 'lose':
                happiness *= -1
            amity_map.add_amity_pair(match.group('foo'),
                                     match.group('bar'),
                                     happiness)

        return amity_map

    def __init__(self):
        """Initialize an empty map."""
        self.names = {}
        self.amity = {}

    def add_amity_pair(self, foo, bar, happiness):
        """Add a pair of people and their (directed) amity to the map."""
        foo_id = self.names.setdefault(foo, len(self.names))
        bar_id = self.names.setdefault(bar, len(self.names))
        self.amity[(foo_id, bar_id)] = happiness

    def add_self(self):
        """Add a neutral (in terms of amity) self to the map."""
        for other in list(self.names):
            self.add_amity_pair('Self', other, 0)
            self.add_amity_pair(other, 'Self', 0)

        # Simplify chaining method calls.
        return self

    def arrangements(self):
        """Yield all possible seating arrangements."""
        for name_list in itertools.permutations(self.names.values()):
            yield name_list, (*name_list[1:], name_list[0])

    def arrangement_happinesses(self):
        """Yield happinesses for all possible seating arrangements."""
        for pair in self.arrangements():
            happiness = 0
            for foo, bar in zip(*pair):
                happiness += self.amity[(foo, bar)] + self.amity[(bar, foo)]
            yield happiness


def main():
    """Main entry point of puzzle solution."""
    amity_map = AmityMap.from_lines(read_input())

    part_one = max(amity_map.arrangement_happinesses())
    part_two = max(amity_map.add_self().arrangement_happinesses())

    print('Part One: {}'.format(part_one))
    print('Part Two: {}'.format(part_two))


if __name__ == '__main__':
    main()
